tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+
+/// Abstracts attribute functionality.
+trait AttributeProvider {
+    fn add_attribute(&mut self, attr: Attribute);
+
+    fn with_attribute(mut self, attr: Attribute) -> Self
+    where
+        Self: Sized,
+    {
+        self.add_attribute(attr);
+        self
+    }
+}
+
+impl CalyxWriter for Attributes {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        for attr in self {
+            attr.write(f)?;
+            f.write_char(' ')?;
+        }
+        Ok(())
+    }
+}
+
+/// A port on a cell or a component.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Port {
+    attributes: Vec<Attribute>,
+    /// The cell that this port belongs to, or `None` if it is in a component
+    /// signature.
+    parent: Option<String>,
+    name: String,
+    /// A width of `-1` means the actual width is unknown/irrelevant/inferred.
+    width: isize,
+}
+
+impl Port {
+    /// Constructs a new port with the given `parent` and `name`.
+    ///
+    /// Requires: `width > 0`.
+    pub fn new<S: ToString, T: ToString>(
+        parent: Option<S>,
+        name: T,
+        width: usize,
+    ) -> Self {
+        assert!(width > 0);
+        Self {
+            attributes: vec![],
+            parent: parent.map(|s| s.to_string()),
+            name: name.to_string(),
+            width: width as isize,
+        }
+    }
+
+    /// Constructs a new port with the given `name` in a component.
+    ///
+    /// Requires: `width > 0`.
+    pub fn new_in_comp<S: ToString>(name: S, width: usize) -> Self {
+        Self::new::<String, S>(None, name, width)
+    }
+
+    /// Constructs a new port with the given `parent` and `name` and with
+    /// inferred width.
+    pub fn inferred<S: ToString, T: ToString>(
+        parent: Option<S>,
+        name: T,
+    ) -> Self {
+        Self {
+            attributes: vec![],
+            parent: parent.map(|s| s.to_string()),
+            name: name.to_string(),
+            width: -1,
+        }
+    }
+
+    pub fn has_inferred_width(&self) -> bool {
+        self.width == -1
+    }
+}
+
+impl AttributeProvider for Port {
+    fn add_attribute(&mut self, attr: Attribute) {
+        self.attributes.push(attr);
+    }
+}
+
+impl Display for Port {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(parent) = &self.parent {
+            write!(f, "{}.", parent)?;
+        }
+        write!(f, "{}", self.name)
+    }
+}
+
+impl CalyxWriter for Port {
+    /// Behaves identically to [`Port::fmt`]. Please read [`CalyxWriter::write`]
+    /// for documentation on this function.
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+/// Abstracts port functionality over components and cells.
+pub trait PortProvider {
+    /// Retrieves the port on this provider named `port`.
+    ///
+    /// Requires: `port` exists on this provider.
+    fn get(&self, port: String) -> Port;
```

Comment:
The `port` argument for this should probably be `AsRef<str>` or at least `&str`.
I don't see a strong reason why we should need to pass an owned string to an
accessor.

Reply:
nit: probably should name this `get_port` for clarity since `get` is a pretty
common method name and using traits to directly name certain methods is not
exactly pretty to look at

Reply:
Ah I see, this is actually a constructor. Feel free to disregard

Reply:
e.g. here I can use refactoring on the `get` method in `PortProvider` to change the name in all of its implementers.